staging/src/k8s.io/apiserver/pkg/util/shufflesharding/shufflesharding.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package shufflesharding
+
+import (
+	"fmt"
+	"math"
+)
+
+// MaxHashBits is the max bit length which can be used from hash value.
+// If we use all bits of hash value, the critical(last) card shuffled by
+// Dealer will be uneven to 2:3 (first half:second half) at most,
+// in order to reduce this unevenness to 32:33, we set MaxHashBits to 60 here.
+const MaxHashBits = 60
+
+// RequiredEntropyBits will figure out the required entropy bits for the given
+// deckSize and handSize, and this value must be less or equal than MaxHashBits
+// if we want to create and use Dealer.
+func RequiredEntropyBits(deckSize, handSize int) int {
+	return int(math.Ceil(math.Log2(float64(deckSize)) * float64(handSize)))
```

Comment:
Remember that floating point arithmetic has limited precision.  `math.Log2(x)` produces only an approximation to the true base-2 logarithm of `x`.  I see no guarantee that this approximation is `<=` the true value.  Thus, the approximation could be slightly higher than the true value.  `math.Ceil` of something slightly higher than an integer is almost a whole integer larger.  That could be a problem here.  I suggest
```
 	return int(math.Ceil(math.Log2(float64(deckSize)) * float64(handSize) - 0.001))
```

Reply:
https://play.golang.org/p/cdiupEIE4EN

It seems it actually has the opposite problem.

Is this a good way to test for this? Would computing a max number be better?

Reply:
That is, that test observed approximations that were low but none that were high.
Looking at the doc on the math package, I see no guarantee in either direction.
The formulation I suggested is safe for our purposes, both for approximations that are a little below the correct value and approximations that are a little above the correct value.

Reply:
@lavalamp @MikeSpreitzer  Thanks for your suggestion, according to https://play.golang.org/p/muwuKH3EWDL, maybe we can use `math/big`, inspiration from @yue9944882 ...

Reply:
Even `math/big` is approximate when dealing with real numbers.  So switching to that changes nothing.

Look carefully at the revision I suggested.  It works when the approximation is low, and it works with the approximation is high.